python/hidet/distributed/store.py

```diff
@@ -0,0 +1,210 @@
+        self._world_size = world_size
+
+        self._lock = filelock.FileLock(self._lock_filename)
+        self._cache = {}
+        self._timeout = None
+
+        num_peers = self._add('cnt', 1)
+        if 0 <= world_size < num_peers:
+            raise RuntimeError("Warning: more peers than world size.")
+
+        # We cannot operate files in __del__, and we don't want to call close explicitly
+        # So we register a atexit function doing cleanup when python interpreter exits
+        @atexit.register
+        def cleanup():
+            with self._lock:
+                if os.path.exists(self._filename):
+                    rest = self._add('cnt', -1)
+                    if rest == 0:
+                        os.remove(self._filename)
+
+    def _write(self, f, content):
+        f.write(struct.pack('i', len(content)))
+        f.write(content)
+
+    def _read(self, f):
+        len_str = f.read(4)
+        if len_str == b'':
+            return None
+        l = struct.unpack('i', len_str)[0]
+        return f.read(l)
+
+    def _file_size(self, f):
+        origin_pos = f.tell()
+        f.seek(0, 2)  # 2 means the file's end
+        size = f.tell()
+        f.seek(origin_pos, 0)
+        return size
+
+    def _update(self, f):
+        self._cache = {}
+        f.seek(0)
+        while True:
+            k = self._read(f)
+            if k is None:
+                return
+            v = self._read(f)
+            k = str(k, encoding='raw_unicode_escape')
+            if k.startswith(self.DELETE_PREFIX):
+                k = k[len(self.DELETE_PREFIX) :]
+                del self._cache[k]
+            else:
+                self._cache[k] = v
+
+    def _add(self, key: str, amount: int) -> int:
+        with self._lock:
+            with open(self._filename, "ab+") as f:
+                f.seek(0)
+                self._update(f)
+            value = int(self._cache.get(key, '0')) + amount
+            with open(self._filename, "ab+") as f:
+                self._write(f, bytes(key, encoding='raw_unicode_escape'))
+                self._write(f, bytes(str(value), encoding='raw_unicode_escape'))
+        return value
+
+    def _check(self, keys: List[str]):
+        with self._lock:
+            with open(self._filename, "ab+") as f:
+                f.seek(0)
+                self._update(f)
+        return all((key in self._cache for key in keys))
+
+    def _set(self, key: str, value: bytes):
+        with self._lock:
+            with open(self._filename, "ab+") as f:
+                self._write(f, bytes(key, encoding='raw_unicode_escape'))
+                self._write(f, value)
+
+    def set(self, key: str, value: bytes) -> None:
+        self._set(self.REGULAR_PREFIX + key, value)
+
+    def get(self, key: str) -> bytes:
+        last_file_size = None
+        key = self.REGULAR_PREFIX + key
+        start_t = datetime.now()
+        while True:
+            self._lock.acquire()
```

Comment:
Seems we can still use `with self._lock` here.

Reply:
Yeah. I forget to remove the comment after I spend some time to understand the design.